Get Upstream Commit reference for people using a local source

karlcow opened the pull request:
#### 83de710ceb2c01600466ced314d13874b4aa6d1b
<pre>
Get Upstream Commit reference for people using a local source
<a href="https://bugs.webkit.org/show_bug.cgi?id=272142">https://bugs.webkit.org/show_bug.cgi?id=272142</a>
<a href="https://rdar.apple.com/125894887">rdar://125894887</a>

Reviewed by Jonathan Bedard.

This provides the hash of the HEAD for people using a local checkout
of WPT tests. This is a followup on webkit.org/b/248445 which gave
this information for people using the remote checkout. We wrapped in
a try/except to make sure the directory was really a checkout of a repo.

It makes it possible for people to directly copy paste in the commit
message the hash of the WPT checkout when they import
or re-import WPT tests.

Example:
--------- Please include the following in your commit message: ---------
Upstream commit: <a href="https://github.com/web-platform-tests/wpt/commit/3324586879a38951098d072bf5ae8f56408a9afa">https://github.com/web-platform-tests/wpt/commit/3324586879a38951098d072bf5ae8f56408a9afa</a>
------------------------------------------------------------------------

* Tools/Scripts/webkitpy/w3c/test_downloader.py:
(TestDownloader.checkout_test_repository):
(TestDownloader):
(TestDownloader.get_rev_parse_head):
* Tools/Scripts/webkitpy/w3c/test_importer.py:
(TestImporter.do_import):

Canonical link: <a href="https://commits.webkit.org/279698@main">https://commits.webkit.org/279698@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a58ce5c68db42793a965f28038f4457bf418926c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54173 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33561 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6711 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57449 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4897 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56476 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41083 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4877 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43871 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3274 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56269 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31806 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46908 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25018 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/54005 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28614 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4228 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3046 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/50320 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4433 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59042 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29376 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/4622 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51291 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30548 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47020 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50647 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11810 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31518 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30328 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/webkitpy/w3c/test_downloader.py:
```diff
@@ -115,7 +115,10 @@ def checkout_test_repository(self, revision, url, directory):
-        self.upstream_revision = git.rev_parse('HEAD')
+        self.upstream_revision = self.get_rev_parse_head(git)
+
+    def get_rev_parse_head(self, git):
```
I don't understand the purpose of this function, instead of just using `git.rev_parse('HEAD')`, since that's all this function does.

karlcow replied:
@JonWBedard it's to be able to use `git.rev_parse` in `test_importer.py` without having to import the full git just getting the hash. So I'm using the one in `test_downloader.py`

karlcow replied:
That said I could probably call it directly too since I have initialized git. Good point.

karlcow replied:
ok fixed. Thanks.

karlcow commented on Tools/Scripts/webkitpy/w3c/test_importer.py:
```diff
@@ -209,7 +209,11 @@ def __init__(self, port, test_paths, options):
     def do_import(self):
-        if not self.source_directory:
+        if self.source_directory:
+            source_path = Path(self.source_directory) / 'web-platform-tests'
```
ha! my use of Pathlib seems incompatible with the tests

in short
```
  TypeError: 'PosixPath' object is not subscriptable
```

In details: 
```
[2029/2630] webkitpy.w3c.test_importer_unittest.TestImporterTest.test_crash_test_with_resource_file erred:
  Traceback (most recent call last):
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_importer_unittest.py", line 470, in test_crash_test_with_resource_file
      fs = self.import_directory(['-s', '/home/user/wpt', '-d', '/mock-checkout/LayoutTests/imported/w3c/web-platform-tests'], FAKE_FILES, ['web-platform-tests/css/css-images'])
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_importer_unittest.py", line 128, in import_directory
      importer.do_import()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_importer.py", line 214, in do_import
      git = self.test_downloader().git(source_path)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_downloader.py", line 102, in git
      return Git(test_repository, None, executive=self._host.executive, filesystem=self._filesystem)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 65, in __init__
      SCM.__init__(self, cwd, **kwargs)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/scm.py", line 68, in __init__
      self.checkout_root = self.find_checkout_root(self.cwd)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 133, in find_checkout_root
      checkout_root = self._filesystem.join(path, checkout_root)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/system/filesystem_mock.py", line 218, in join
      if comp[0] == sep:
  TypeError: 'PosixPath' object is not subscriptable
  
```

karlcow replied:
fixed.

karlcow commented on Tools/Scripts/webkitpy/w3c/test_importer.py:
```diff
@@ -209,7 +209,12 @@ def __init__(self, port, test_paths, options):
+        if self.source_directory:
+            # We need to convert the path as a string to make the tests framework happy
+            source_path = str(Path(self.source_directory) / 'web-platform-tests')
+            git = self.test_downloader().git(source_path)
```
hmm so I guess my two remaining failues is because I bring the full test_downloader in a scenario which was not expected for the tests cases before. 

```
OpenSource % /usr/bin/python3 ./Tools/Scripts/test-webkitpy               
Suppressing most webkitpy logging while running unit tests.
Skipping tests in the following modules or packages because they are really, really, slow:
    webkitpy.common.checkout.scm.scm_unittest
    (https://bugs.webkit.org/show_bug.cgi?id=31818; use --all to include)

[1/2630] reporelaypy.tests.checkoutroute_unittest.RedirectorUnittest.test_fallback passedMock process 44483 send signal 9
Mock process 44496 send signal 9
Mock process 44496 send signal 9
[25/2630] webkitbugspy.tests.bugzilla_unittest.TestBugzilla.test_comments passedMock process 44484 send signal 9
Mock process 44488 send signal 9
[393/2630] webkitpy.common.net.bugzilla.bugzilla_unittest.BugzillaQueriesTest.test_load_query passed/Users/karlcow/code/OpenSource/Tools/Scripts/libraries/autoinstalled/python-3-arm64/bs4/builder/__init__.py:545: XMLParsedAsHTMLWarning: It looks like you're parsing an XML document using an HTML parser. If this really is an HTML document (maybe it's XHTML?), you can ignore or filter this warning. If it's XML, you should know that using an XML parser will be more reliable. To parse this document as XML, make sure you have the lxml package installed, and pass the keyword argument `features="xml"` into the BeautifulSoup constructor.
  warnings.warn(
[874/2630] webkitpy.layout_tests.run_webkit_tests_integrationtest.RunTest.test_single_file passedProcess Process-8:4:                                       
[1483/2630] webkitpy.layout_tests.run_webkit_tests_integrationtest.RunTest.test_skipped_flag (+9)worker/1 starting
worker/0 starting
worker/0 stopping
worker/1 stopping
[1776/2630] webkitpy.w3c.test_importer_unittest.TestImporterTest.test_import_dir_with_no_tests_and_no_hg erred:                           
  Traceback (most recent call last):
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/memoized.py", line 42, in __call__
      return self._results_cache[args]
  KeyError: (<webkitpy.port.config.Config object at 0x10c071fa0>,)
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_importer_unittest.py", line 85, in test_import_dir_with_no_tests_and_no_hg
      importer.do_import()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_importer.py", line 215, in do_import
      git = self.test_downloader().git(source_path)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_downloader.py", line 102, in git
      return Git(test_repository, None, executive=self._host.executive, filesystem=self._filesystem)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 65, in __init__
      SCM.__init__(self, cwd, **kwargs)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/scm.py", line 68, in __init__
      self.checkout_root = self.find_checkout_root(self.cwd)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 131, in find_checkout_root
      checkout_root = self._run_git(['rev-parse', '--show-toplevel'], cwd=(path or "./")).strip()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 110, in _run_git
      return self.run(full_command_args, **full_kwargs)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/scm.py", line 73, in run
      return self._executive.run_command(args,
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/system/executive_mock.py", line 230, in run_command
      raise self._exception  # pylint: disable=E0702
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/port/config.py", line 123, in _read_configuration
      configuration_path = self._filesystem.join(self.build_directory(None), "Configuration")
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/port/config.py", line 78, in build_directory
      output = self._executive.run_command(args, cwd=self._webkit_finder.webkit_base(), return_stderr=False).rstrip()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/system/executive_mock.py", line 230, in run_command
      raise self._exception  # pylint: disable=E0702
  OSError
  
[2103/2630] webkitscmpy.test.git_unittest.TestGit.test_files_changed passedMock process 44493 send signal 9                               
[2140/2630] webkitpy.w3c.test_importer_unittest.TestImporterTest.test_import_dir_with_no_tests erred:
  Traceback (most recent call last):
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/memoized.py", line 42, in __call__
      return self._results_cache[args]
  KeyError: (<webkitpy.port.config.Config object at 0x134252550>,)
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_importer_unittest.py", line 97, in test_import_dir_with_no_tests
      importer.do_import()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_importer.py", line 215, in do_import
      git = self.test_downloader().git(source_path)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/w3c/test_downloader.py", line 102, in git
      return Git(test_repository, None, executive=self._host.executive, filesystem=self._filesystem)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 65, in __init__
      SCM.__init__(self, cwd, **kwargs)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/scm.py", line 68, in __init__
      self.checkout_root = self.find_checkout_root(self.cwd)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 131, in find_checkout_root
      checkout_root = self._run_git(['rev-parse', '--show-toplevel'], cwd=(path or "./")).strip()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/git.py", line 110, in _run_git
      return self.run(full_command_args, **full_kwargs)
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/checkout/scm/scm.py", line 73, in run
      return self._executive.run_command(args,
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/system/executive_mock.py", line 230, in run_command
      raise self._exception  # pylint: disable=E0702
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/port/config.py", line 123, in _read_configuration
      configuration_path = self._filesystem.join(self.build_directory(None), "Configuration")
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/port/config.py", line 78, in build_directory
      output = self._executive.run_command(args, cwd=self._webkit_finder.webkit_base(), return_stderr=False).rstrip()
    File "/Users/karlcow/code/OpenSource/Tools/Scripts/webkitpy/common/system/executive_mock.py", line 230, in run_command
      raise self._exception  # pylint: disable=E0702
  webkitpy.common.system.executive.ScriptError: abort: no repository found in '/Volumes/Source/src/wk/Tools/Scripts/webkitpy/w3c' (.hg not found)!
  
Ran 2630 tests in 20.969s                                                                                                                     
FAILED (failures=0, errors=2)
```

karlcow replied:
This test https://github.com/WebKit/WebKit/blob/45c4d855298d4aa4bdb43aefb0ba8b69af92f888/Tools/Scripts/webkitpy/w3c/test_importer_unittest.py#L74-L86

And this test
https://github.com/WebKit/WebKit/blob/45c4d855298d4aa4bdb43aefb0ba8b69af92f888/Tools/Scripts/webkitpy/w3c/test_importer_unittest.py#L87-L97

karlcow replied:
The two tests are here to check when calling `do_import()` with empty directory and with empty directory and no checkout. 

my modification switches from a model where it was 

```diff
-        if not self.source_directory:
+        if self.source_directory:
+            …
+        else:
```

karlcow replied:
unittests are harder to write when the function, here `do_import`, does too many things.

karlcow replied:
ok cool. I wrapped the in a try/except clause. Because indeed if the source directory is not a checkout, it would fail.

karlcow committed:
Get Upstream Commit reference for people using a local source
https://bugs.webkit.org/show_bug.cgi?id=272142
rdar://125894887

Reviewed by Jonathan Bedard.

This provides the hash of the HEAD for people using a local checkout
of WPT tests. This is a followup on webkit.org/b/248445 which gave
this information for people using the remote checkout. We wrapped in
a try/except to make sure the directory was really a checkout of a repo.

It makes it possible for people to directly copy paste in the commit
message the hash of the WPT checkout when they import
or re-import WPT tests.

Example:
--------- Please include the following in your commit message: ---------
Upstream commit: https://github.com/web-platform-tests/wpt/commit/3324586879a38951098d072bf5ae8f56408a9afa
------------------------------------------------------------------------

* Tools/Scripts/webkitpy/w3c/test_downloader.py:
(TestDownloader.checkout_test_repository):
(TestDownloader):
(TestDownloader.get_rev_parse_head):
* Tools/Scripts/webkitpy/w3c/test_importer.py:
(TestImporter.do_import):

Canonical link: https://commits.webkit.org/279698@main
